### PR TITLE
`ray.experimental.serve` Module

### PR DESCRIPTION
**Pull request by simon-mo**
This PR proposes a new module name `ray.serve`

`ray.serve` is a module for publishing your actors to interact with outside world. It utilizes ray to horizontally scale your actors. 

### Architecture
In the following illustration, call chain goes from top to bottom. 
Each box is one or more replicated ray actors.

 ```
             +-------------------+     +-----------------+   +------------+
Frontend     |   HTTP Frontend   |     |    Arrow RPC    |   |    ...     |
  Tier       |                   |     |                 |   |            |
             +-------------------+     +-----------------+   +------------+
              +------------------------------------------------------------+
                   +--------------------+        +-------------------+
 Router           |   Default Router   |        |   Deadline Aware  |
  Tier            |                    |        |      Router       |
                  +--------------------+        +-------------------+
              +------------------------------------------------------------+
                  +----------------+   +--------------+    +-------------+
 Managed         |  Managed Actor |   |     ...      |    |     ...     |
 Actor           |    Replica     |   |              |    |             |
 Tier            +----------------+   +--------------+    +-------------+
```

 ### Frontend Tier
The frontend tier is repsonsible for interface with the world. Currently `ray.serve` will provide
implementation for HTTP Frontend and a zeromq high performance frontend. 

 ### Router Tier
The router tier receives calls from frontend and route them to the managed actors. Routers both _route_ and _queue_ incoming queries. `ray.serve` has native support for (micro-)batching queries. 

In addition, we implemented a deadline aware routers that will put high priority queries in the front
of the queue so they will be delivered first. 

 ### Managed Actor Tier
Managed actors will be managed by routers. These actors can contains arbitrary methods. Methods in the actors class are assumed to be able to take into a batch of input at a time. If this cannot be assumed, you can use the `@single_input` decorator, `ray.serve` will run your method in a for loop working on the micro-batch. 
   
cc @atumanov @robertnishihara 

P.S.
This is a re-implementation of an [earlier prototype](https://github.com/simon-mo/corvette)

### CODE REVIEW

**Review by robertnishihara — COMMENTED** [review]
Awesome stuff! Left some comments/questions.

File: python/ray/serve/tests/frontend/test_default_app.py [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import ray
```
[the remark this answers is not in the document]
Comment (robertnishihara) [reply python/ray/serve/tests/frontend/test_default_app.py]: though note that our travis tests currently uses python 3.5

**Comment by jovany-wang** [comment]
My concern is that the `python` http server impemented by `Starlette` may not meet the perf requirements for such real-time scenario.

**Comment by simon-mo** [comment]
Combined with unicorn does have better performance than tornado. That’s why
we are considering it.

https://www.techempower.com/benchmarks/#section=data-r17&hw=ph&test=fortune&l=zijzen-1

**Comment by jovany-wang** [comment]
@simon-mo Thanks for your reply.
I'd like to test the perf of  uvicorn and some other C++ http framework.

File: python/ray/experimental/serve/README.md [hunk 2]
```diff
@@ -0,0 +1,63 @@
+# Ray Serve Module
```
Comment (robertnishihara) [inline python/ray/experimental/serve/README.md]: This file is redundant.

File: python/ray/experimental/serve/README.rst [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+priority queries in the front of the queue so they will be delivered
+first.
+
+Managed Actor Tier
+~~~~~~~~~~~~~~~~~~
+
+Managed actors will be managed by routers. These actors can contains
+arbitrary methods. Methods in the actors class are assumed to be able to
+take into a batch of input at a time. If this cannot be assumed, you can
+use the ``@single_input`` decorator, it will run your method in a for
```
Comment (robertnishihara) [inline python/ray/experimental/serve/README.rst]: Change single input to default instead of batch.

**Comment by robertnishihara** [comment]
When I run `python -m pytest -v -s python/ray/experimental/serve/tests/`, the test `frontend/test_default_app.py::test_http_basic` seems to hang. It succeeds when I run the test individually. Any idea about this?

**Comment by robertnishihara** [comment]
I also see this failure

```
python/ray/experimental/serve/tests/router/test_deadline_router.py::test_deadline_priority FAILED

================================================== FAILURES ==================================================
___________________________________________ test_deadline_priority ___________________________________________

router = Actor(DeadlineAwareRouter, cc9b8191335004e9e1ef56e110e4b5fbaf9f3dc9), now = 405663.805075376

    def test_deadline_priority(router: DeadlineAwareRouter, now: float):
        # first sleep 2 seconds
        first = unwrap(router.call.remote("SleepCounter", 2, now + 1))
    
        # then send a request to with deadline farther away
        second = unwrap(router.call.remote("SleepCounter", 0, now + 10))
    
        # and a request with sooner deadline
        third = unwrap(router.call.remote("SleepCounter", 0, now + 1))
    
        id_1, id_2, id_3 = ray.get([first, second, third])
    
>       assert id_1 < id_3 < id_2
E       assert 3 < 2

python/ray/experimental/serve/tests/router/test_deadline_router.py:88: AssertionError
===================================== 1 failed, 4 passed in 7.15 seconds =====================================
```

**Comment by simon-mo** [comment]
@robertnishihara can we just call the tests separately? I have been debugging this for some time now and still can't fix it somehow, the only clue it got is that:
```
2019-03-08 00:49:58,104	WARNING actor.py:652 -- Actor is garbage collected in the wrong driver. Actor id = ActorID(2cd834f87ff272c3a9a34ff74fe83d33ef001877), class name = ScalerAdder.
```

The ordering issue was due to oversubscribtion, should be fixed now.

**Review by robertnishihara — APPROVED** [review]
Thanks @simon-mo, this looks great! I'll merge this once the CI passes.

As discussed offline, we should treat `uvicorn` and `starlette` as placeholders for the time being and not make any architectural decisions that tie us to them.

**Comment by robertnishihara** [comment]
We should probably also run these tests in our CI, but that will require upgrading to 3.6 since starlette requires 3.6. That can be done in a follow up PR.

**Comment by pcmoritz** [comment]
Can we rename ray.serve to ray.serving in some future PR? That seems more natural for me!

**Pull request merged**